src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -312,124 +309,82 @@ void FileBasedWal::prepareNewFile(LogID startLogId) {
+                       << errno << "): " << strerror(errno);
+            close(fd);
+            return 0;
+        }
 
-    // Rollover if required
-    if (buffer->needToRollover()) {
-        closeCurrFile();
-    }
+        if (id == logId) {
+            // Found
+            close(fd);
+            return term;
+        }
 
-    Cord cord;
-    LogID firstIdInCord = buffer->firstLogId();
-    auto accessFn = [&cord, &firstIdInCord, this] (
-            LogID id,
-            TermID term,
-            ClusterID cluster,
-            const std::string& log) {
-        cord << id << term << int32_t(log.size()) << cluster;
-        cord.write(log.data(), log.size());
-        cord << int32_t(log.size());
-
-        size_t currSize = currFd_ >= 0 ? currInfo_->size() : 0;
-        if (currSize + cord.size() > maxFileSize_) {
-            dumpCord(cord, firstIdInCord, id, term);
-            // Reset the cord
-            cord.clear();
-            firstIdInCord = id + 1;
-
-            // Need to close the current file and create a new file
-            closeCurrFile();
-        }
-    };
-
-    // Dump the buffer to file
-    auto lastLog = buffer->accessAllLogs(accessFn);
-
-    // Dump the rest if any
-    if (!cord.empty()) {
-        dumpCord(cord, firstIdInCord, lastLog.first, lastLog.second);
-    }
+        // Read the message length
+        int32_t len;
+        if (pread(fd, &len, sizeof(int32_t), pos + sizeof(LogID) + sizeof(TermID))
+                != sizeof(int32_t)) {
+            LOG(ERROR) << "Failed to read the message length (errno "
+                       << errno << "): " << strerror(errno);
+            close(fd);
+            return 0;
+        }
 
-    // Flush the wal file
-    if (currFd_ >= 0) {
-        CHECK_EQ(fsync(currFd_), 0);
+        // Move to the next log
+        pos += sizeof(LogID)
+               + sizeof(TermID)
+               + sizeof(ClusterID)
+               + 2 * sizeof(int32_t)
+               + len;
     }
 
-    // Remove the buffer from the list
-    {
-        std::lock_guard<std::mutex> g(buffersMutex_);
-
-        CHECK_EQ(buffer.get(), buffers_.front().get());
-        buffers_.pop_front();
-    }
-    slotReadyCV_.notify_one();
+    LOG(FATAL) << "Should never reach here";
 }
 
 
-BufferPtr FileBasedWal::createNewBuffer(
-        LogID firstId,
-        std::unique_lock<std::mutex>& guard) {
-    if (buffers_.size() >= policy_.numBuffers) {
-        // Log appending is way too fast
-        LOG(WARNING) << "Write buffer is exhausted,"
-                        " need to wait for vacancy";
-        // TODO: Output a counter here
-        // Need to wait for a vacant slot
-        slotReadyCV_.wait(guard, [self = shared_from_this()] {
-            return self->buffers_.size() < self->policy_.numBuffers;
-        });
-    }
+BufferPtr FileBasedWal::getLastBuffer(LogID id, size_t expectedToWrite) {
+    std::unique_lock<std::mutex> g(buffersMutex_);
+    if (!buffers_.empty()) {
```

Comment:
Why we still need a list of buffers?   
One buffer is enough now.

Reply:
Totally agreed.

Since the current implementation already implemented buffer list, I just reused it. In order to keep the PR small, I suggest to switch to the ring buffer in a separate PR. What do you think?